Limit the array index of FixedHashTable by min/max

jiebinn opened the pull request:
We have observed the serial mergeSingleLevel time of Query 7 in ClickBench has costed a lot. Diving into the performance issue, we have found most of the extra cycles have been spent on the `isZero()` loop condition in iterator ++ of fixedHashTable. This is a patch that fix the performance issue if the key range has only occupied small part of the total 16 bits in fixedHashTable.

![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/002336af-8f35-48b6-8fa6-c7639dd32d92)


If the type of key is 8 bits or 16 bits in aggregation, ClickHouse will use array of 256 or 65536 length to store the key and boost the mergeSingleLevel, rather than key comparison. However, if the key has occupied only small range of the total 65536 cells, most of the cycles are wasted on the `isZero()` to find the next cell which is not zero in iterator++.

The solution is to use min/max and update min/max when emplace. Then we can set the upper searching limit to max in iterator++. And just set min as the value of `begin()`, rather than searching the first cell that not equals to 0.

We have tested the patch on 2x80 vCPUs server, Query 7 of ClickBench has gained 2.1x performance improvement. There is no regression for the other queries of ClickBench. The overall geomean has got 2% performance improvement.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Speed up FixedHashTable by keeping track of the min and max keys used. This allows to reduce the number of cells that need to be verified.


### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ] <!---ci_include_fast--> Fast test
- [ ] <!---ci_include_integration--> Integration Tests
- [ ] <!---ci_include_stateless--> Stateless tests
- [ ] <!---ci_include_stateful--> Stateful tests
- [ ] <!---ci_include_unit--> Unit tests
- [ ] <!---ci_include_performance--> Performance tests
- [ ] <!---ci_include_asan--> All with ASAN
- [ ] <!---ci_include_tsan--> All with TSAN
- [ ] <!---ci_include_analyzer--> All with Analyzer
- [ ] <!---ci_include_KEYWORD--> Add your option here

#### Exclude tests:
- [ ] <!---ci_exclude_fast--> Fast test
- [ ] <!---ci_exclude_integration--> Integration Tests
- [ ] <!---ci_exclude_stateless--> Stateless tests
- [ ] <!---ci_exclude_stateful--> Stateful tests
- [ ] <!---ci_exclude_performance--> Performance tests
- [ ] <!---ci_exclude_asan--> All with ASAN
- [ ] <!---ci_exclude_tsan--> All with TSAN
- [ ] <!---ci_exclude_msan--> All with MSAN
- [ ] <!---ci_exclude_ubsan--> All with UBSAN
- [ ] <!---ci_exclude_coverage--> All with Coverage
- [ ] <!---ci_exclude_aarch64--> All with Aarch64
- [ ] <!---ci_exclude_KEYWORD--> Add your option here

#### Extra options:
- [ ] <!---do_not_test--> do not test (only style check)
- [ ] <!---no_merge_commit--> disable merge-commit (no merge from master before tests)
- [ ] <!---no_ci_cache--> disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ] <!---batch_0--> 1
- [ ] <!---batch_1--> 2
- [ ] <!---batch_2--> 3
- [ ] <!---batch_3--> 4

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #62746 from jiebinn/ClickHouse:FixedHashTable -->
*This is an automated comment for commit 588162647633449fb34c7f010e2c4b7f9539d78a with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62746/588162647633449fb34c7f010e2c4b7f9539d78a/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

jiebinn commented:
Hi @nickitat, thanks to review this PR. BTW, could you help to add the "can be tested" label?

nickitat commented on src/Common/HashTable/FixedHashTable.h:
```diff
@@ -294,36 +296,28 @@ class FixedHashTable : private boost::noncopyable, protected Allocator, protecte
     const_iterator begin() const
     {
-        if (!buf)
+        if (!buf && min > max)
```
pls let's clarify what `min > max` means.
maybe it makes sense to extract it into a separate function with a self-explanatory name. afaiu it means that container is empty.

jiebinn replied:
Yes. In that case, `min > max` means the container is empty. Then outside `it` is always larger than the `end()` and the for loop will not stop.

```
    void ALWAYS_INLINE mergeToViaEmplace(Self & that, Func && func)
    {
        for (auto it = this->begin(), end = this->end(); it != end; ++it)
        {
            typename Self::LookupResult res_it;
            bool inserted;
            that.emplace(it->getKey(), res_it, inserted, it.getHash());
            func(res_it->getMapped(), it->getMapped(), inserted);
        }
    }
```

nickitat commented on src/Common/HashTable/FixedHashTable.h:
```diff
@@ -294,36 +296,28 @@ class FixedHashTable : private boost::noncopyable, protected Allocator, protecte
     const_iterator begin() const
     {
-        if (!buf)
+        if (!buf && min > max)
```
maybe it should be `!buf || min > max`?

jiebinn replied:
Sorry for the late reply.
Yes, it should be `!buf || min > max`. Thanks to correct that.

jiebinn committed:
Limit the array index of FixedHashTable by min/max

If the type of key is 8 bits or 16 bits in aggregation, ClickHouse will use array
of 256 or 65536 length to store the key and boost the mergeSingleLevel, rather than
key comparison. However, if the key has occupied only small range of the total 65536
cells, most of the cycles are wasted on the `isZero()` to find the next cell which
is not zero in iterator++.
The solution is to use min/max and update min/max when emplace. Then we can set the
upper searching limit to max in iterator++. And just set min as the value of `begin()`,
rather than searching the first cell that not equals to 0.
We have tested the patch on 2x80 vCPUs server, Query 7 of ClickBench has gained 2.1x
performance improvement.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Fix a bug if the container is empty

jiebinn commented:
I'm working on the testing failures.

jiebinn committed:
Fix a bug if data will be inserted not by emplace().

jiebinn committed:
Add the use_emplace_to_insert_data flag. `emplace()` is the only interface
to update min/max. If the FixedHashTable.emplace() is not used to revise
the hashtable value, then we should not continue the min/max optimization.

jiebinn commented:
Hi @nickitat , sorry for the late reply as the holiday and personal's leave for one week. Here is the latest update.
I have added the flag `use_min_max_optimization` to decide if we should this optimization or fallback. There are two cases that we should fall back to the original code path and not use the min/max optimization.
1. The FixedHashTable is empty and min/max is not set by the `emplace()`. Then `min > max` in this case.
2. The `data()` of FixedHashTable is used to insert raw columns into the FixedHashTable, not by `emplace()`. `emplace()` is the only Interface to update min/max. Then the boundary(min/max) is invalid in this case.

I have checked the current testing failure. It's caused by the Client connect issue and not related to the PR. The tests are passed locally.

jiebinn commented:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{mso-number-format:"0\.0%";}
.xl66
	{text-align:right;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Query: | OPT/BASE
-- | --
0 | 105.8%
1 | 100.2%
2 | 100.2%
3 | 99.3%
4 | 99.5%
5 | 100.1%
6 | 101.4%
7 | 184.3%
8 | 104.1%
9 | 103.9%
10 | 102.4%
11 | 103.5%
12 | 101.9%
13 | 105.4%
14 | 103.1%
15 | 103.0%
16 | 103.9%
17 | 106.8%
18 | 111.3%
19 | 100.2%
20 | 99.9%
21 | 100.0%
22 | 100.3%
23 | 100.1%
24 | 99.3%
25 | 99.1%
26 | 99.8%
27 | 100.3%
28 | 99.0%
29 | 102.1%
30 | 101.6%
31 | 102.7%
32 | 101.3%
33 | 98.8%
34 | 106.0%
35 | 100.9%
36 | 99.9%
37 | 100.6%
38 | 99.7%
39 | 100.2%
40 | 100.1%
41 | 100.5%
42 | 100.5%
Overall | 103.0%



</body>

</html>

nickitat commented on src/Common/HashTable/FixedHashTable.h:
```diff
@@ -114,6 +114,9 @@ template <typename Key, typename Cell, typename Size, typename Allocator>
 class FixedHashTable : private boost::noncopyable, protected Allocator, protected Cell::State, protected Size
 {
     static constexpr size_t NUM_CELLS = 1ULL << (sizeof(Key) * 8);
+    bool use_emplace_to_insert_data = true;
+    size_t min = NUM_CELLS - 1;
+    size_t max = 0;
```
```suggestion
    /// We maintain min and max values inserted into the hash table to then limit the amount of cells to traverse to the [min; max] range.
    /// Both values could be efficiently calculated only within `emplace` calls (and not when we populate the hash table in `read` method for example), so we update them only within `emplace` and track if any other method was called.
    bool only_emplace_was_used_to_insert_data = true;
    size_t min = NUM_CELLS - 1;
    size_t max = 0;
```

nickitat commented on src/Common/HashTable/FixedHashTable.h:
```diff
@@ -298,9 +303,14 @@ class FixedHashTable : private boost::noncopyable, protected Allocator, protecte
-        auto buf_end = buf + NUM_CELLS;
-        while (ptr < buf_end && ptr->isZero(*this))
-            ++ptr;
+        if (!use_min_max_optimization())
```
I think it also makes sense to abstract 

``` cpp
return use_min_max_optimization() ? but + min : buf;
```

and the same for `end` into a separate function. so all other code won't know anything inside this logic, it will just call `firstPopulatedCell()` .

jiebinn replied:
That's great. Thanks for your kind suggestion. The latest commit has packed the related code into `firstPopulatedCell()` and `lastPopulatedCell()`.

nickitat commented on src/Common/HashTable/FixedHashTable.h:
```diff
@@ -377,6 +401,10 @@ class FixedHashTable : private boost::noncopyable, protected Allocator, protecte
 
+    /// Decide if we use the min/max optimization. `max < min` means the FixedHashtable is empty. The flag `use_emplace_to_insert_data`
+    /// will check if the FixedHashTable will use `emplace()` to insert the raw data.
+    bool ALWAYS_INLINE use_min_max_optimization() const {return ((max >= min) && use_emplace_to_insert_data);}
```
```suggestion
    bool ALWAYS_INLINE canUseMinMaxOptimization() const {return ((max >= min) && use_emplace_to_insert_data);}
```

jiebinn committed:
Update src/Common/HashTable/FixedHashTable.h

Add comment by Nikita.

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

jiebinn committed:
Update src/Common/HashTable/FixedHashTable.h

Revise the method name by Nikita.

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

nickitat commented:
`Stateful tests (ubsan)` - failure looks related

jiebinn committed:
Generate the seperate function firstPopulatedCell() and lastPopulatedCell()

jiebinn commented:
Hi @nickitat, I have found that lots of recent PRs also produced the same msan/tsan error. I have checked the error log and the error is produced while building none x86 binary. I don't think that is related.
[#63942]https://github.com/ClickHouse/ClickHouse/pull/63985
[#63942]https://github.com/ClickHouse/ClickHouse/pull/63942
[#63985]https://github.com/ClickHouse/ClickHouse/pull/63985

azat commented:
You need to rebase, sanitizers had been fixed in https://github.com/ClickHouse/ClickHouse/pull/64090

jiebinn commented:
> You need to rebase, sanitizers had been fixed in #64090

Thanks!

nickitat commented:
> Hi @nickitat, I have found that lots of recent PRs also produced the same msan/tsan error. I have checked the error log and the error is produced while building none x86 binary. I don't think that is related. [#63942] [#63942] [#63985]

report has complains about the new code https://s3.amazonaws.com/clickhouse-test-reports/62746/627722c285b610a5499e705bf1725af2621fccc1/stateful_tests__ubsan_/stderr.log

jiebinn committed:
Avoid UBSan warning while buf is nullptr

jiebinn commented:
Rebase the code and fix the ubsan warning. It seems the current binary (`binary_riscv64`, `binary_loongarch64`) compiling error is not related and also existed in PR #64175, #64128, #64202.

jiebinn commented:
Hi @nickitat, thanks to review the PR. I'm trying to fix the CI failure by rebase the master branch and check that. Currently, there are two types of testing failures.
1. By the network connection error.
2. By the existing reported issue #64308.
Do you think what else should I do to help the PR?

nickitat commented:
> Hi @nickitat, thanks to review the PR. I'm trying to fix the CI failure by rebase the master branch and check that. Currently, there are two types of testing failures.
> 
> 1. By the network connection error.
> 2. By the existing reported issue [Broken upgrade check: New settings are not reflected in settings changes history #64308](https://github.com/ClickHouse/ClickHouse/issues/64308).
>    Do you think what else should I do to help the PR?

I think we're fine. thanks for patience

jiebinn commented:
> > Hi @nickitat, thanks to review the PR. I'm trying to fix the CI failure by rebase the master branch and check that. Currently, there are two types of testing failures.
> > 
> > 1. By the network connection error.
> > 2. By the existing reported issue [Broken upgrade check: New settings are not reflected in settings changes history #64308](https://github.com/ClickHouse/ClickHouse/issues/64308).
> >    Do you think what else should I do to help the PR?
> 
> I think we're fine. thanks for patience

Thanks to review the PR!
